### PR TITLE
Add standardized compiler flags for GNU and Intel with enhanced debugging options

### PR DESCRIPTION
**Pull request by amstokely**
This PR introduces standardized Fortran compiler flags tailored for the GNU and Intel compilers, enhancing both build consistency and debugging capability across all build types. The updates are as follows:

- **GNU Compiler Flags**:
  - **General**: `-ffree-line-length-none` (removes line length limits), and `-mcmodel=medium` (supports larger memory models for larger datasets).
  - **Debug**: Aggressive debugging flags including `-fbacktrace` (provides error backtrace), `-ffpe-trap=invalid,zero,overflow` (traps floating-point exceptions), and `-fcheck=all` (enables all runtime checks).

- **Intel Compiler Flags**:
  - **General**: `-mcmodel=medium`  (supports larger memory models for larger datasets.)
  - **Debug**: Includes `-check uninit` (checks uninitialized variables), `-ftrapuv` (traps uninitialized variable usage), `-g` (produces debug information), `-traceback` (provides symbolic traceback on errors), and `-fpe0` (halts on floating-point exceptions).

These flags standardize the compilation environment, making it easier for developers to catch issues, particularly in Debug builds where aggressive error-trapping is employed.

### CODE REVIEW

**Comment by ibanos90** [comment]
Hi @amstokely, these are great additions! I wonder if we should update the instructions about how to build the code using these changes. Thanks!

**Comment by amstokely** [comment]
> Hi @amstokely, these are great additions! I wonder if we should update the instructions about how to build the code using these changes. Thanks!
I can update the README with instructions on how to compile in Debug/Release mode.

File: cmake/Obs2Ioda_CompilerFlags.cmake [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+set(FORTRAN_COMPILER_GNU_FLAGS
+    $<$<COMPILE_LANGUAGE:Fortran>:-cpp -ffree-line-length-none -mcmodel=medium>
+)
+
+# Set Debugging Fortran compiler flags specific to the GNU Compiler
+# -fbacktrace: Provide a backtrace when an error occurs
+# -ffpe-trap=invalid,zero,overflow: Trap floating point exceptions (invalid calculation, divide by zero, overflow)
+# -fcheck=all: Execute all types of runtime checks
+set(FORTRAN_COMPILER_GNU_DEBUG_FLAGS
+    $<$<COMPILE_LANGUAGE:Fortran>:-fbacktrace -ffpe-trap=invalid,zero,overflow -fcheck=all>
```
Comment (mgduda) [inline cmake/Obs2Ioda_CompilerFlags.cmake:14]: Is `-g` needed here as well?

**Comment by mgduda** [comment]
@amstokely If I recall correctly the Fortran 2003 standard specifies a maximum of 132 columns for free-format source. Is there a specific reason for adding the `-ffree-line-length-none` flag to not enforce the standard limit?

Also, is pre-processing currently required for any Fortran source, and if not, is there a specific reason for adding `-cpp` / `-fpp`? An alternative would be to rename source files to use the `.F90` suffix, which almost universally causes Fortran compilers to automatically pre-process the source. This might be preferable, as it wouldn't require adding compiler-specific flags for every Fortran compiler that we aim to support.

File: cmake/Obs2Ioda_CompilerFlags.cmake (hunk 1 above)
Comment (amstokely) [reply cmake/Obs2Ioda_CompilerFlags.cmake:14]: Generally CMake will add -g when the build type is set to Debug, but I will go ahead and add -g if this changes in future versions of CMake. Thanks for pointing this out!

**Comment by amstokely** [comment]
> @amstokely If I recall correctly the Fortran 2003 standard specifies a maximum of 132 columns for free-format source. Is there a specific reason for adding the `-ffree-line-length-none` flag to not enforce the standard limit?
> 
> Also, is pre-processing currently required for any Fortran source, and if not, is there a specific reason for adding `-cpp` / `-fpp`? An alternative would be to rename source files to use the `.F90` suffix, which almost universally causes Fortran compilers to automatically pre-process the source. This might be preferable, as it wouldn't require adding compiler-specific flags for every Fortran compiler that we aim to support.

@mgduda -ffree-line-length-none was included in the original Makefile, which is why I included it here. I can remove this flag and see if it causes it any problems. That is good to know about using .F90 vs .f90 for enabling the preprocessor. I'll remove these flags for now and if I need to use the preprocessor in the future I'll try changing the file ext to .F90.

File: cmake/Obs2Ioda_CompilerFlags.cmake (hunk 1 above)
Comment (mgduda) [reply cmake/Obs2Ioda_CompilerFlags.cmake:14]: I suspected that CMake might add `-g` for debug builds, but then I noticed that `-g` was being explicitly added to `FORTRAN_COMPILER_INTEL_DEBUG_FLAGS` as part of this PR.

File: obs2ioda-v2/README.md [hunk 2]
```diff
@@ -22,9 +22,9 @@ If you have an environment preconfigured for `mpas-jedi`, simply source that env
    ```bash
    find . -name *libbufr*
    ```
-4. Next, run CMake to configure the build. Remember to specify the path to the NCEP BUFR library:
+4. Run CMake to configure the build next. Specify the `CMAKE_BUILD_TYPE` option to set the build type. Currently, the supported types are `Release`, `RelWithDebInfo`, and `Debug`. Don't forget to include the path to the NCEP BUFR library:
```
Comment (mgduda) [inline obs2ioda-v2/README.md:25]: The change to the first sentence here seems less grammatically correct to me.

**Review by ibanos90 — APPROVED** [review]
I tested this branch with amsua and ahi observations and it worked great! It's very nice to have these debugging options when adding new code. Thanks @amstokely!

**Review by junmeiban — APPROVED** [review]
I successfully compiled the code using the 'debug' option as mentioned in the README and successfully converted the conventional OBS and AMSU-A/MHS data. Thank you so much @amstokely !

**Pull request merged**